### PR TITLE
Assets settings cleanup

### PR DESCRIPTION
**Pull request by victorbenichoux**
In this PR I clean up a lot of the settings and initialization mechanisms of AssetsManager as inspired by https://github.com/clustree/modelkit/pull/86

The idea is to avoid having the initialization logic split between the validation of settings, and the initialization of the objects themselves. To do so, I do away with using pydantic's BaseSettings classes, and resort to resolving environment variables manually (and explicitly) in the inits.

This allows us to have explicit errors, namely, when a storage provider is requested (via an env var), if it fails to load an error is raised. Otherwise, we keep the AssetsManager without a storage provider (as expected). In addition, it is now possible to initialize the StorageProvider in a more explicit way.

In addition, I also rename the remote storage provider to a more legible name, improve coverage and rework tests.

Important renames:
- `s/AssetsManager.remote_store/AssetsManager.storage_provider`
- `s/RemoteAssetsStore/StorageProvider`
- `s/ModelLibrary.asset_manager/ModelLibrary.assets_manager`

### CODE REVIEW

File: modelkit/assets/drivers/abc.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import abc
+from typing import Iterator
+
+
+class StorageDriver(abc.ABC):
```
Comment (antoinejeannot) [inline modelkit/assets/drivers/abc.py:5]: 🙌

File: modelkit/assets/drivers/gcs.py [hunk 2]
```diff
@@ -1,41 +1,37 @@
...
 
-    class Config:
-        extra = "forbid"
+    def __init__(
+        self,
+        bucket: str = None,
+        service_account_path: str = None,
+        client: Client = None,
+    ):
+        self.bucket = bucket or os.environ.get("MODELKIT_STORAGE_BUCKET", None)
```
Comment (antoinejeannot) [inline modelkit/assets/drivers/gcs.py:26]: NB: ```python self.bucket = bucket or os.environ.get("MODELKIT_STORAGE_BUCKET")```

File: modelkit/assets/drivers/gcs.py [hunk 3]
```diff
@@ -1,41 +1,37 @@
...
+    bucket: storage
+    client: Client
 
-    class Config:
-        extra = "forbid"
+    def __init__(
+        self,
+        bucket: str = None,
+        service_account_path: str = None,
+        client: Client = None,
```
Comment (antoinejeannot) [inline modelkit/assets/drivers/gcs.py:24]: ```python        client: Optional[Client] = None,```

File: modelkit/assets/drivers/s3.py [hunk 4]
```diff
@@ -1,45 +1,49 @@
...
+class S3StorageDriver(StorageDriver):
+    bucket: str
+    endpoint_url: Optional[str]
+    client: Any
 
-    class Config:
-        extra = "forbid"
+    def __init__(
+        self,
+        bucket: str = None,
```
Comment (antoinejeannot) [inline modelkit/assets/drivers/s3.py:23]: Optionals, everywhere

File: modelkit/assets/drivers/s3.py [hunk 5]
```diff
@@ -1,45 +1,49 @@
...
+    ):
 
-
-class S3StorageDriver:
-    def __init__(self, settings: S3DriverSettings = None):
-        if not settings:
-            settings = S3DriverSettings()
-        self.bucket = settings.bucket
-        self.endpoint_url = settings.s3_endpoint
+        self.bucket = bucket or os.environ.get("MODELKIT_STORAGE_BUCKET", None)
```
Comment (antoinejeannot) [inline modelkit/assets/drivers/s3.py]: NB: ```python         self.bucket = bucket or os.environ.get("MODELKIT_STORAGE_BUCKET")```

Comment (antoinejeannot) [reply modelkit/assets/drivers/s3.py]: ^ same for every `get` :D

File: modelkit/assets/manager.py [hunk 6]
```diff
@@ -218,7 +224,7 @@ def fetch_asset(
         self,
         spec: Union[AssetSpec, str],
         return_info=False,
-        force_download: Optional[bool] = None,
+        force_download: bool = None,
```
Comment (antoinejeannot) [inline modelkit/assets/manager.py:227]: Too bad for the Optional

File: modelkit/assets/remote.py [hunk 7]
```diff
@@ -32,15 +34,50 @@ def get_size(dir_path):
...
+        **driver_settings,
+    ):
+        self.timeout = timeout_s or int(
+            os.environ.get("MODELKIT_STORAGE_TIMEOUT_S", 300)
+        )
+        self.prefix = (
+            prefix or os.environ.get("MODELKIT_STORAGE_PREFIX") or "modelkit-assets"
+        )
+        self.force_download = force_download or bool(
+            os.environ.get("MODELKIT_STORAGE_FORCE_DOWNLOAD", False)
```
Comment (antoinejeannot) [inline modelkit/assets/remote.py:66]: NB: no need for `False`

File: tests/assets/test_drivers.py [hunk 8]
```diff
@@ -41,3 +43,22 @@ def test_gcs_driver(gcs_assetsmanager):
...
 def test_s3_driver(s3_assetsmanager):
     _perform_driver_test(s3_assetsmanager.storage_provider.driver)
+
+
+def test_local_driver_overwrite(working_dir):
+    driver = LocalStorageDriver(working_dir)
+    driver.upload_object(
+        os.path.join(TEST_DIR, "assets", "testdata", "some_data.json"), "a/b/c"
+    )
+    # will remove the a/b/c file
```
Comment (antoinejeannot) [inline tests/assets/test_drivers.py:53]: we also might need to check what is going on on disk, in addition to these comments

**Review by antoinejeannot — APPROVED** [review]
Despite few minor comments, great work !

File: modelkit/assets/manager.py [hunk 9]
```diff
@@ -37,28 +37,32 @@ def _has_succeeded(local_path):
...
-        self.assets_dir = settings.assets_dir
-        self.timeout = settings.timeout
-
-        self.storage_provider = None
-        if settings.remote_store:
+    def __init__(
+        self,
+        assets_dir: str = None,
+        timeout: int = None,
+        storage_provider: StorageProvider = None,
```
[the remark this answers is not in the document]
Comment (victorbenichoux) [reply modelkit/assets/manager.py:48]: I added Optional types to all keyword arguments, as suggested in https://www.python.org/dev/peps/pep-0483/#pragmatics

File: modelkit/assets/manager.py [hunk 10]
```diff
@@ -37,28 +37,32 @@ def _has_succeeded(local_path):
...
+        self.storage_provider = storage_provider
+        if not self.storage_provider:
             try:
-                self.storage_provider = StorageProvider(**settings.remote_store.dict())
+                self.storage_provider = StorageProvider()
                 logger.debug(
                     "AssetsManager created with remote storage provider",
                     driver=self.storage_provider.driver,
                 )
             except BaseException:
```
[the remark this answers is not in the document]
Comment (victorbenichoux) [reply modelkit/assets/manager.py]: In later commits, I use the `NoConfiguredProviderError` raised by the `StorageProvider`'s init, and leave all others flow through

File: modelkit/assets/drivers/gcs.py (hunk 2 above)
Comment (victorbenichoux) [reply modelkit/assets/drivers/gcs.py:26]: Thanks, done

File: modelkit/assets/drivers/gcs.py (hunk 3 above)
Comment (victorbenichoux) [reply modelkit/assets/drivers/gcs.py:24]: fixed

File: modelkit/assets/drivers/s3.py (hunk 4 above)
Comment (victorbenichoux) [reply modelkit/assets/drivers/s3.py:23]: 😄

File: modelkit/assets/drivers/s3.py (hunk 5 above)
Comment (victorbenichoux) [reply modelkit/assets/drivers/s3.py]: Thanks, I also fixed it here, and aligned the bucket init to other drivers

File: modelkit/assets/manager.py (hunk 6 above)
Comment (victorbenichoux) [reply modelkit/assets/manager.py:227]: the one good Optional. I really thought they were no longer  necessary, although it is the other way around

File: modelkit/assets/remote.py (hunk 7 above)
Comment (victorbenichoux) [reply modelkit/assets/remote.py:66]: fixed

**Comment by victorbenichoux** [comment]
thanks for the review @antoinejeannot , PTAL

**Review by antoinejeannot — APPROVED** [review]

**Comment by antoinejeannot** [comment]
GG 🔥

**Pull request merged**